### DISCUSSION TITLE
Lambda to function pointer

### DISCUSSION THREAD

**Discussion by cincuranet** [dotnet/roslyn#55847] [Q&A]
I'm manually building lambda (`Expression.Lambda`, `xxx.Compile()`, etc.) to have a specific code for a given set of dynamically provided parameters. The result is always `Func<A, B, TResult>`. I there a way to "cast" this to function pointer? I'm trying to avoid the virtual call from `Func<A, B, TResult>.Invoke`.

**Comment by PathogenDavid** [comment]
TL;DR: You can, but you *really* shouldn't.

You cannot safely do this because the implementation details of lambdas is not specified. In general though, you can expect them to be instance methods on a display class. ([For example, look at what the compiler generates here](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTCDMACBFANgKTwBUBTAZwBcBhAGwmuswG9M9uD9CSEAVgA8sWgBo8YyWIB8FGrQCylWgAsA9gBMAFAEouPIwF55OiJOB68pvBDwBqPMADcmAL5A=).) Function pointers [cannot point to instance methods](https://github.com/dotnet/csharplang/blob/b89d4c934041db923f7238b1427cd5f3ae71ed4b/proposals/csharp-9.0/function-pointers.md#allow-instance-methods), therefore it is not safe to call a lambda via function pointer.

------------

That being said, if you're feeling spicy you *can* access the function pointer like this:

```csharp
Func<int, int, int> add = (a, b) => a + b;
delegate* managed<object, int, int, int> addP = (delegate* managed<object, int, int, int>)add.Method.MethodHandle.GetFunctionPointer();
Console.WriteLine($"2 + 7 = {addP(null, 2, 7)}");
```

The `object` parameter is actually the `this` parameter for the display class. It can be `null` in this case since the lambda does not capture any variables. Again, **this is relying on implementation details of both Roslyn and the runtime**, I would not recommend doing it outside of testing for novelty's sake.

-----------

> I'm trying to avoid the virtual call from `Func<A, B, TResult>.Invoke`.

This isn't *really* avoiding the virtual call. A virtual call *is* a call via function pointer.

That being said, calling directly via a function pointer as shown above is *sometmes*\* marginally faster if the function pointer is stored in a static readonly field and JIT is able to optimize away the field access:

|                                 Method |      Mean |     Error |    StdDev | Ratio | RatioSD | Code Size | Allocated |
|--------------------------------------- |----------:|----------:|----------:|------:|--------:|----------:|----------:|
|          ViaLambda_StaticReadonlyField | 1.0983 ns | 0.0241 ns | 0.0213 ns |  1.00 |    0.00 |      35 B |         - |
| ViaFunctionPointer_StaticReadonlyField | 0.8111 ns | 0.0198 ns | 0.0185 ns |  0.74 |    0.02 |      31 B |         - |
|                ViaLambda_InstanceField | 1.1296 ns | 0.0308 ns | 0.0288 ns |  1.03 |    0.04 |      26 B |         - |
|       ViaFunctionPointer_InstanceField | 0.9177 ns | 0.0454 ns | 0.0987 ns |  0.91 |    0.13 |      20 B |         - |

\*Sometimes because it's such tiny overhead that sometimes it comes out to be the same, although generally `ViaFunctionPointer_StaticReadonlyField` benchmarks slightly faster on my machine. The reality is, the body of your lambda will probably end up dwarfing the time spend on calling it.

<details>
<summary>(Click to show benchmark code)</summary>

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;

IConfig config = DefaultConfig.Instance;
config = config.AddDiagnoser(MemoryDiagnoser.Default);
config = config.AddDiagnoser(new DisassemblyDiagnoser(new(exportHtml: true, printSource: true, exportCombinedDisassemblyReport: true)));
BenchmarkRunner.Run<TestBenchmark>(config);

public unsafe class TestBenchmark
{
    private static readonly Func<int, int, int> Lambda_StaticReadonlyField;
    private static readonly delegate*<object, int, int, int> FunctionPointer_StaticReadonlyField;

    public Func<int, int, int> Lambda_InstanceField;
    public delegate*<object, int, int, int> FunctionPointer_InstanceField;

    static TestBenchmark()
    {
        Lambda_StaticReadonlyField = (a, b) => a + b;
        FunctionPointer_StaticReadonlyField = (delegate*<object, int, int, int>)Lambda_StaticReadonlyField.Method.MethodHandle.GetFunctionPointer();
    }

    public TestBenchmark()
    {
        Lambda_InstanceField = Lambda_StaticReadonlyField;
        FunctionPointer_InstanceField = FunctionPointer_StaticReadonlyField;
    }

    [Benchmark(Baseline = true)]
    public int ViaLambda_StaticReadonlyField()
        => Lambda_StaticReadonlyField(3226, 42);

    [Benchmark]
    public int ViaFunctionPointer_StaticReadonlyField()
        => FunctionPointer_StaticReadonlyField(null!, 3226, 42);

    [Benchmark]
    public int ViaLambda_InstanceField()
        => Lambda_InstanceField(3226, 42);

    [Benchmark]
    public int ViaFunctionPointer_InstanceField()
        => FunctionPointer_InstanceField(null!, 3226, 42);
}
```

</details>

**Reply by cincuranet** [reply]
Sadly the `Method.MethodHandle.GetFunctionPointer` does not work for `DynamicMethod` (which is what I have from all the `Expression.Xxx`).

Regarding the call, I want to avoid going through `callvirt` and have `calli` used. Which the `delegate*` should result in.
